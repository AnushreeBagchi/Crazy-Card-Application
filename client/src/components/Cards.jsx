import React from "react";
import "../../app/index.css";
import { connect } from "react-redux";
import CrazyCard from "./CrazyCard.jsx";
import {
  onCardSelected,
  resetSelectedCards,
  onCardRemoved,
  fetchCards,
  fetchCardDetails,
  loadSelectedCards,
} from "../store/actions/card";
import Button from "@material-ui/core/Button";
import SelectGrid from "./SelectGrid";
import { HOME_BTN_TEXT,LIQUID_INCOME_RANGE, CARD_DETAILS } from "../constants/constants";

class Cards extends React.Component {
  async componentDidMount() {
    await this.props.fetchCards();
    this.props.loadSelectedCards();
  }

  getAvailableCards(customer) {
    if (this.props.state.card.cards) {
      let available = ["Anywhere Card"];
      if (customer.income >= LIQUID_INCOME_RANGE) {
        available.push("Liquid Card");
      }
      if (customer.empStatus === "Student") {
        available.push("Student Life Card");
      }
      return this.props.state.card.cards.filter((card) =>
        available.includes(card.name)
      );
    }
  }

  goToHome() {
    this.props.resetSelectedCards();
    this.props.history.push({
      pathname: "/",
    });
  }

  render() {
    let customer = this.props.location.state
      ? this.props.location.state.customer
      : this.goToHome();
    let availableCards = customer ? this.getAvailableCards(customer) : null;
    return (
      <>
        <div className="page">
          <div className="card-div">
            {availableCards &&
              availableCards.map((card) => (
                <CrazyCard
                  key={card.name}
                  card={card}
                  cardDetails={CARD_DETAILS}
                  onSelect={(card) => this.props.onCardSelected(card)}
                  onRemove={(card) => this.props.onCardRemoved(card)}
                />
              ))}
          </div>
          {this.props.state.card.selectedCards && (
            <SelectGrid selectedCards={this.props.state.card.selectedCards} />
          )}
          <Button
            size="medium"
            color="primary"
            className="home-btn flex-end"
            onClick={() => this.goToHome()}
          >
            {HOME_BTN_TEXT}
          </Button>
        </div>
      </>
    );
  }
}

const mapStateToProps = (state) => ({ state: state });
const mapDispatchToProps = (dispatch) => ({
  onCardSelected: (card) => {
    dispatch(onCardSelected(card));
  },
  onCardRemoved: (card) => {
    dispatch(onCardRemoved(card));
  },
  resetSelectedCards: (card) => {
    dispatch(resetSelectedCards(card));
  },
  fetchCards: () => {
    dispatch(fetchCards());
  },
  fetchCardDetails: () => {
    dispatch(fetchCardDetails());
  },
  loadSelectedCards: () => {
    dispatch(loadSelectedCards());
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(Cards);
