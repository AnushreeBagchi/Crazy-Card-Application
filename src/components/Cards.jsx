import React from "react";
import "../../app/index.css";
import Card from "@material-ui/core/Card";
import CardActionArea from "@material-ui/core/CardActionArea";
import CardActions from "@material-ui/core/CardActions";
import CardContent from "@material-ui/core/CardContent";
import CardMedia from "@material-ui/core/CardMedia";
import Button from "@material-ui/core/Button";
import Typography from "@material-ui/core/Typography";

class Cards extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      cards: [
        {
          name: "Student Life Card",
          image: "../src/images/cards/studentLife.png",
          apr: "18.9%",
          balanceOfferDuration: "0 months",
          purchaseOfferDuration: "6 months",
          credit: "£1200",
          
        },
        {
          name: "Anywhere Card",
          image: "../src/images/cards/anywhere.png",
          apr: "33.9%",
          balanceOfferDuration: "0 months",
          purchaseOfferDuration: "0 months",
          credit: "£300",
        },
        {
          name: "Liquid Card",
          image: "../src/images/cards/liquid.png",
          apr: "33.9%",
          balanceOfferDuration: "12 months",
          purchaseOfferDuration: "6 months",
          credit: "£3000",
          minIncome : 16000
        }
      ],
      cardDetails: [
        {
          name: "apr",
          text: "Representative % APR (variable)",
        },
        {
          text: "Balance Transfer Offer Duration",
          name: "balanceOfferDuration",
        },
        {
          text: "Purchase Offer Duration",
          name: "purchaseOfferDuration",
        },
        {
          text: "Credit Available",
          name: "credit",
        },
      ],
    };
  }
  render() {
    return (
      <>
        {this.state.cards.map((card) => (
          <Card key={card.name} className="card">
            <Typography gutterBottom variant="h5" component="h2">
              {card.name}
            </Typography>
            <div className="flex media-div">
            <CardMedia className="media" image={card.image} />
            <CardActionArea>
              <CardContent className="card-content">
                {this.state.cardDetails.map((detail) => (
                  <div key={detail.name} className="content">
                    <Typography variant="body2" color="textSecondary" component="p" >
                      {detail.text}
                    </Typography>
                    <Typography className="detail-value" variant="h6" color="textSecondary" component="p" >
                      {card[detail.name]}
                    </Typography>
                  </div>
                ))}
              </CardContent>
            </CardActionArea>
            </div>
            <CardActions>
              <Button size="small" color="primary">
                Learn More
              </Button>
            </CardActions>
          </Card>
        ))}
      </>
    );
  }
}

export default Cards;
