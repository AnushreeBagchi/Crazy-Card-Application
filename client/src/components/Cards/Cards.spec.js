import React from "react";
import { shallow } from "enzyme";
import { Cards } from "./Cards.jsx";

describe("Tests for Cards", () => {
  const fetchCards = () => ({});
  const fetchAvailableCards = () => ({});
  const loadSelectedCards = () => ({});
  const state = {
    card: [
      {
        name: "Anywhere Card",
        image: "https://ddialhyn49dxu.cloudfront.net/cards/mbna2019-large.png",
        apr: "33.9%",
        balanceOfferDuration: "0 months",
        purchaseOfferDuration: "0 months",
        credit: "Â£300",
        min_income: 0,
        empStatus: null,
      },
    ],
  };

  it("Renders correctly", () => {
    const wrapper = shallow(
      <Cards
        state={state}
        fetchCards={fetchCards}
        fetchAvailableCards={fetchAvailableCards}
        loadSelectedCards={loadSelectedCards}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
