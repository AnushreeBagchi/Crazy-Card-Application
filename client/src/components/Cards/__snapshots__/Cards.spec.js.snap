// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for Card Renders correctly 1`] = `
<WithStyles(ForwardRef(Container))>
  <WithStyles(ForwardRef(Grid))
    container={true}
    direction="row"
    spacing={4}
  >
    <WithStyles(ForwardRef(Grid))
      item={true}
      lg={8}
      md={6}
      sm={12}
      xs={12}
    >
      <WithStyles(ForwardRef(Grid))
        container={true}
        direction="row"
        spacing={4}
      />
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      item={true}
      lg={4}
      md={6}
      sm={12}
      xs={12}
    />
  </WithStyles(ForwardRef(Grid))>
</WithStyles(ForwardRef(Container))>
`;
